asdf/commands/tests/test_edit.py

```diff
@@ -0,0 +1,231 @@
+import os
+import shutil
+
+import numpy as np
+import pytest
+
+import asdf
+from asdf import AsdfFile
+from asdf.commands import main
+from ...tests.helpers import get_file_sizes, assert_tree_match
+
+
+
+"""
+Three tests are defined.  
+
+1. Run the command 'asdftool edit -e' to create a YAML file, simulating
+   the steps a user would make to start editing an ASDF file.
+2. Run the command 'asdftool edit -s' to save YAML edits such that the
+   edited YAML will have the same or fewer characters as the original
+   ASDF file, so will be overwritten in place with the space character
+   used as any buffer, to consume all the memory on disk the YAML takes
+   up in the ASDF file.
+2. Run the command 'asdftool edit -s' to save YAML edits such that the
+   edited YAML will have the more characters than the original ASDF file.
+   This triggers a rewrite of the file, since there isn't enough 'room' 
+   on disk to accomadate the edited YAML.  The resultant YAML will be 
+   rewritten with a buffer (using the space character as buffer) to 
+   accomodate future edits.  If a block index existed in the original ASDF
+   file, it will need to be recomputed and if one didn't exist, it will be
+   added to the resultant ASDF file.  
+"""
+
+
+
+def create_base_asdf(tmpdir, version):
+    """
+    In the test temp directory, create a base ASDF file to edit
+    and test against.
+    """
+    seq = np.arange(100)
+
+    # Store the data in an arbitrarily nested dictionary
+    tree = {
+        "foo": 42,
+        "name": "Monty",
+        "sequence": seq,
+    }
+
+    fname = "test_edit_base.asdf"
+    oname = os.path.join(tmpdir, fname)
+    if os.path.exists(oname):
+        os.remove(oname)
+    af = asdf.AsdfFile(tree,version=version)
+    af.write_to(oname)
+
+    return oname
+
+
+def create_edit_equal(base_yaml):
```

Comment:
These three helpers could be combined into a single function, something like this:

```python
import re

def create_edited_yaml(input, output, pattern, replacement):
    with open(input) as fd:
        content = fd.read()
    new_content = re.sub(pattern, replacement)
    with open(output, "w") as fd:
      fd.write(new_content)
```

Then in the test:

```python
# Edit with equal length
create_edited_yaml(yaml_base, edited_path, "foo: 41", "foo: 42")

# Edit with shorter length
create_edited_yaml(yaml_base, edited_path, "foo: 41", "foo: 2")

# Edit with longer length
create_edited_yaml(yaml_base, edited_path, "foo: 41", "foo: 41\nbar: 13")
```

Reply:
I refactored the creation of the edited YAML using the `re` module.